packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,657 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
+
+        self.points.splice(startIndex, self.points.length - startIndex)
+        self.points.splice(0, endIndex)
+        self.points.splice(self.points.length, 0, ...self.splicePoints)
       } else {
-        /*
-        Segment spans the start point of a closed loop, so
-        trim the ends of the line back to dragIndex and targetIndex.
-        */
-        self.trim(startIndex, endIndex)
+        for (var i = startIndex; i < endIndex; i++) {
+          closeSpliceAction.pointsFromDrag.push(self.points[i])
+        }
+
+        if (self.spliceReverse) {
+          self.splicePoints = self.splicePoints.reverse()
+        }
+
+        self.points.splice(startIndex, endIndex - startIndex, ...self.splicePoints)
+      }
+
+      self.spliceReset()
+      self.setDragPoint(self.pathIsClosed ? null : self.points.at(-1))
+      self.setClosePoint(self.pathIsClosed ? null : self.points.at(0))
+
+      return closeSpliceAction
+    },
+
+    undo() {
+      if (self.drawActions.length == 0) {
+        // remove a mark that was created, not loaded
+        if (self.points.length == 1) {
+          self.finish()
+        }
+        return
+      }
+      let action = toJS(self.drawActions.at(-1))
+
+      if (action.type == 'start') {
+        self.drawActions.pop()
+        self.tool.deleteMark(self)
+      } else if (action.type == 'end') {
+        self.points.pop()
+        self.setDragPoint(self.points.at(-1))
+        self.setClosePoint(self.points.at(0))
+        self.drawActions.pop()
+      } else if (action.type == 'append') {
+        let indexStart = self.points.length - action.pointsFromDrag.length
+        self.points.splice(indexStart, action.pointsFromDrag.length)
+        self.setDragPoint(self.points.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-drag-point') {
+        self.spliceDragPointIndex = null
+        self.setDragPoint((self.pathIsClosed) ? null : self.points.at(-1))
+        self.setClosePoint((self.pathIsClosed) ? null : self.points.at(0))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-close-point') {
+        self.splicePoints.length = 0
+        self.setClosePoint(null)
+        self.spliceReverse = false
+        self.spliceThroughBeginning = false
+        self.spliceActive = false
+        self.spliceClosePointIndex = null
+        self.drawActions.pop()
+      } else if (action.type == 'splice-append') {
+        let indexStart = self.splicePoints.length - action.pointsFromDrag.length
+        self.splicePoints.splice(indexStart, action.pointsFromDrag.length)
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-end') {
+        if (action.throughBeginning) {
+          let spliceIndex = Math.abs(action.pointIndexDrag - action.pointIndexClose)
+
+          for (let i = spliceIndex; i < self.points.length - 1; i++) {
+            self.splicePoints.push(self.points[i])
+          }
```

Comment:
Pushing points one by one is slow here. Can you copy the required points from `self.points` into a new array, then map or concat that array onto `self.splicePoints`?
```suggestion
          const splicedPoints = self.points.slice(spliceIndex)
          self.splicePoints.concat(splicedPoints)
```

Reply:
Something to be aware of with MobX State Tree is that it's opinionated about storage. There can only be one instance of any tree node, and it will error if you try to store the same thing in two places. So `self.splicePoints` might need to store references to points, rather than actual points themselves. Otherwise, I can see a couple of potential problems when you work with real lines with thousands of points:

1. MST crashes whenever you try to copy points, rather than pass references.
2. Your undo and redo stacks contain thousands of copies of the original line points, using up memory and making the code much slower.

There's an `UndoManager` middleware for MST that stores patch diffs for each change, rather than the objects themselves, which is much, much faster. You store a patch for each new action, then either reverse that patch for undo or re-apply the patch for redo. It works by recording patches with [`recordPatches`](https://mobx-state-tree.js.org/API/#recordpatches).
https://github.com/mobxjs/mobx-state-tree/blob/50e6df954da72fd2875bffb14bda0d59e134c792/packages/mst-middlewares/src/undo-manager.ts#L68-L85

Reply:
This was a huge pain in my butt at various points and why I opted for so much volatile state. The uniqueness requirement for points made undo/redo with array manipulation particularly painful. It likely could be solved by creating individually unique points whenever there's an array operation but that sounds inefficient and unnecessarily painful for what is ultimately temporary state management.

Reply:
>Which was just messy and clunky with each point having to be its own object in its transition from an undo state to a redo state to an applied state. I wish it had worked because it would have made this code a bit cleaner.

I ran into this same problem when I was working on freehand lines last year. Storing line state in two places, like you've got here (`mark.pathX` and `mark.pathY` for the stored annotation, `mark.points` for the rendered line) was just messy and hard to manage. FWIW, what I found worked was to derive `mark.points` from `mark.pathX` and mark.pathY`, with any changes to points happening in the original observable arrays. So `mark.pathX` and `mark.pathY` are the source of truth for line paths, and `mark.points` is a derived array that's convenient when you need to work with the path as a single array. This view made the line state a lot easier to manage:
https://github.com/zooniverse/front-end-monorepo/blob/ad85fb308c5c0e0b0d2d05455360f7487f39d987/packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js#L60-L65